File: old-sepa-sources-rejected/python/migrate_sepa_sources.py
"""Find WooCommerce subscriptions still charging a legacy Stripe SEPA Source and
migrate them to a supported SEPA Debit PaymentMethod before the next renewal fails.

Stripe stopped accepting old `src_...` Sources for off-session SEPA renewals. An
order or subscription that still points at one of these will be rejected at the next
charge, usually with an error like "This payment method is not supported for this
API version" or the Source status shows "canceled" or "chargeable=false". This walks
recent renewal orders, reads the saved token from `_stripe_intent_id` / transaction_id
or order meta, and for any legacy Source, looks up the Stripe Customer for a newer
SEPA Debit PaymentMethod that can replace it. When one exists it re-links the order
and the subscription's saved token; otherwise it flags the order for the shopper to
re-enter their IBAN. Safe by default (DRY_RUN=true). Run on a schedule.
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("migrate_sepa_sources")

stripe.api_key = os.environ.get("STRIPE_SECRET_KEY", "sk_test_dummy")
WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"), os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"))
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "30"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

RENEWAL_STATUSES = {"pending", "on-hold", "failed"}
LEGACY_SOURCE_PREFIX = "src_"
SEPA_PM_TYPE = "sepa_debit"


def token_of(order):
    """The saved Stripe token for this order, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid or None


def is_legacy_source(token):
    return bool(token) and token.startswith(LEGACY_SOURCE_PREFIX)


def decide(order, token, replacement_pm):
    """Pure decision: what should we do about this order's saved payment token.

    order: dict with at least "status" and "id".
    token: the saved Stripe token string, or None.
    replacement_pm: a SEPA Debit PaymentMethod id (str) found on the customer, or None.

    Returns a tuple (action, reason) where action is one of:
      "skip"    - nothing to do, not a legacy Source problem
      "migrate" - a modern SEPA PaymentMethod exists, relink it
      "flag"    - legacy Source with no replacement, ask the shopper to re-enter their IBAN
    """
    if order["status"] not in RENEWAL_STATUSES:
        return ("skip", "order is not awaiting or retrying a renewal")
    if not is_legacy_source(token):
        return ("skip", "saved token is not a legacy Source")
    if replacement_pm:
        return ("migrate", "legacy Source found, a SEPA Debit PaymentMethod is available")
    return ("flag", "legacy Source found, no SEPA Debit PaymentMethod on file")


def find_sepa_payment_method(customer_id):
    """Return the id of the newest reusable SEPA Debit PaymentMethod for this customer, if any."""
    if not customer_id:
        return None
    methods = stripe.PaymentMethod.list(customer=customer_id, type=SEPA_PM_TYPE, limit=10)
    if not methods.data:
        return None
    newest = max(methods.data, key=lambda pm: pm.created)
    return newest.id


def renewal_orders():
    page = 1
    after = f"{__import__('datetime').date.today() - __import__('datetime').timedelta(days=LOOKBACK_DAYS)}T00:00:00"
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={
                "status": "pending,on-hold,failed",
                "after": after,
                "per_page": 50,
                "page": page,
            },
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def stripe_customer_id_of(order):
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_customer_id" and meta.get("value"):
            return meta["value"]
    return None


def migrate(order, replacement_pm):
    """Relink the order (and its parent subscription, if any) to the new PaymentMethod."""
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}",
        json={
            "meta_data": [
                {"key": "_stripe_intent_id", "value": replacement_pm},
                {"key": "_stripe_source_id", "value": replacement_pm},
            ]
        },
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={
            "note": (
                "Migrated from a legacy Stripe SEPA Source to PaymentMethod "
                f"{replacement_pm}. This order can now be retried or will use "
                "the new token on the next renewal."
            )
        },
        auth=AUTH, timeout=30,
    ).raise_for_status()


def flag(order, reason):
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={
            "note": (
                f"SEPA payment check failed: {reason}. This order is on a legacy "
                "Stripe Source that Stripe no longer accepts for renewals, and no "
                "replacement SEPA Debit PaymentMethod was found. The shopper needs "
                "to re-enter their IBAN on the account page before the next renewal."
            )
        },
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    migrated = 0
    flagged = 0
    for order in renewal_orders():
        token = token_of(order)
        customer_id = stripe_customer_id_of(order)
        replacement_pm = find_sepa_payment_method(customer_id) if is_legacy_source(token) else None
        action, reason = decide(order, token, replacement_pm)
        if action == "skip":
            continue
        log.info(
            "Order %s: %s. %s",
            order["id"], reason,
            "would " + action if DRY_RUN else action + "ing",
        )
        if action == "migrate":
            if not DRY_RUN:
                migrate(order, replacement_pm)
            migrated += 1
        elif action == "flag":
            if not DRY_RUN:
                flag(order, reason)
            flagged += 1
    log.info(
        "Done. %d order(s) %s, %d order(s) %s.",
        migrated, "to migrate" if DRY_RUN else "migrated",
        flagged, "to flag" if DRY_RUN else "flagged",
    )


if __name__ == "__main__":
    run()

File: old-sepa-sources-rejected/node/migrate-sepa-sources.js
/**
 * Find WooCommerce subscriptions still charging a legacy Stripe SEPA Source and
 * migrate them to a supported SEPA Debit PaymentMethod before the next renewal fails.
 *
 * Stripe stopped accepting old `src_...` Sources for off-session SEPA renewals. An
 * order or subscription that still points at one of these will be rejected at the
 * next charge. This walks recent renewal orders, reads the saved token, and for any
 * legacy Source, looks up the Stripe Customer for a newer SEPA Debit PaymentMethod
 * that can replace it. When one exists it re-links the order; otherwise it flags the
 * order for the shopper to re-enter their IBAN. Safe by default (DRY_RUN=true).
 *
 * Guide: https://www.allanninal.dev/woocommerce/old-sepa-sources-rejected/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 30);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const RENEWAL_STATUSES = new Set(["pending", "on-hold", "failed"]);
const LEGACY_SOURCE_PREFIX = "src_";
const SEPA_PM_TYPE = "sepa_debit";

export function tokenOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  return order.transaction_id || null;
}

export function isLegacySource(token) {
  return Boolean(token) && token.startsWith(LEGACY_SOURCE_PREFIX);
}

/**
 * Pure decision: what should we do about this order's saved payment token.
 *
 * order: object with at least { status, id }.
 * token: the saved Stripe token string, or null.
 * replacementPm: a SEPA Debit PaymentMethod id (string) found on the customer, or null.
 *
 * Returns [action, reason] where action is one of "skip", "migrate", "flag".
 */
export function decide(order, token, replacementPm) {
  if (!RENEWAL_STATUSES.has(order.status)) {
    return ["skip", "order is not awaiting or retrying a renewal"];
  }
  if (!isLegacySource(token)) {
    return ["skip", "saved token is not a legacy Source"];
  }
  if (replacementPm) {
    return ["migrate", "legacy Source found, a SEPA Debit PaymentMethod is available"];
  }
  return ["flag", "legacy Source found, no SEPA Debit PaymentMethod on file"];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function findSepaPaymentMethod(customerId) {
  if (!customerId) return null;
  const methods = await stripe.paymentMethods.list({ customer: customerId, type: SEPA_PM_TYPE, limit: 10 });
  if (!methods.data.length) return null;
  const newest = methods.data.reduce((a, b) => (a.created > b.created ? a : b));
  return newest.id;
}

function stripeCustomerIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_customer_id" && meta.value) return meta.value;
  }
  return null;
}

async function* renewalOrders() {
  const after = new Date(Date.now() - LOOKBACK_DAYS * 86400000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(`/orders?status=pending,on-hold,failed&after=${after}&per_page=50&page=${page}`);
    if (!batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

async function migrate(order, replacementPm) {
  await woo(`/orders/${order.id}`, {
    method: "PUT",
    body: JSON.stringify({
      meta_data: [
        { key: "_stripe_intent_id", value: replacementPm },
        { key: "_stripe_source_id", value: replacementPm },
      ],
    }),
  });
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Migrated from a legacy Stripe SEPA Source to PaymentMethod ${replacementPm}. ` +
            `This order can now be retried or will use the new token on the next renewal.`,
    }),
  });
}

async function flag(order, reason) {
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `SEPA payment check failed: ${reason}. This order is on a legacy Stripe Source ` +
            `that Stripe no longer accepts for renewals, and no replacement SEPA Debit ` +
            `PaymentMethod was found. The shopper needs to re-enter their IBAN on the account page.`,
    }),
  });
}

export async function run() {
  let migrated = 0;
  let flagged = 0;
  for await (const order of renewalOrders()) {
    const token = tokenOf(order);
    const customerId = stripeCustomerIdOf(order);
    const replacementPm = isLegacySource(token) ? await findSepaPaymentMethod(customerId) : null;
    const [action, reason] = decide(order, token, replacementPm);
    if (action === "skip") continue;
    console.log(`Order ${order.id}: ${reason}. ${DRY_RUN ? "would " + action : action + "ing"}`);
    if (action === "migrate") {
      if (!DRY_RUN) await migrate(order, replacementPm);
      migrated++;
    } else if (action === "flag") {
      if (!DRY_RUN) await flag(order, reason);
      flagged++;
    }
  }
  console.log(
    `Done. ${migrated} order(s) ${DRY_RUN ? "to migrate" : "migrated"}, ` +
    `${flagged} order(s) ${DRY_RUN ? "to flag" : "flagged"}.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}
